numpy/core/src/multiarray/getset.c

```diff
@@ -529,11 +529,30 @@ array_descr_set(PyArrayObject *self, PyObject *arg, void *NPY_UNUSED(ignored))
             axis = 0;
         }
         else {
-            /* Don't mention the deprecated F-contiguous support */
-            PyErr_SetString(PyExc_ValueError,
-                    "To change to a dtype of a different size, the array must "
-                    "be C-contiguous");
-            goto fail;
+            if (newtype->elsize < PyArray_DESCR(self)->elsize) {
+                // For smaller elements, check if there is a unit axis
+                for(axis = PyArray_NDIM(self) - 1; axis >= 0; axis--) {
```

Comment:
The loop could be eliminated for a single check with `axis = PyArray_NDIM(self) - 1;`.

Reply:
I don't think the version with the loop is correct, as it means the semantics of `view` are affected by the underlying memory layout of the array. My impression is that the behavior should be that the semantics of `view` are independent of the underlying memory representation, but an exception is raised if the representation doesn't support the semantics.